separate out error code in panel

braver opened the pull request:
Error codes, especially when they're more texty than your average pythonesque numbered code, kinda get lost in the panel sometimes.

Before:
<img width="816" alt="screen shot 2018-08-11 at 08 52 41" src="https://user-images.githubusercontent.com/2543659/43989112-edf549c6-9d43-11e8-8c91-6f772a83c261.png">

After:
<img width="783" alt="screen shot 2018-08-11 at 08 52 08" src="https://user-images.githubusercontent.com/2543659/43989115-f349431e-9d43-11e8-8f88-b6a8494d6519.png">

kaste commented:
Subtle. I would have to customize my scheme to like it.

Before: 

![image](https://user-images.githubusercontent.com/8558/43999731-8d5e6244-9e12-11e8-8426-e1b5224f45ff.png)


After:

![image](https://user-images.githubusercontent.com/8558/43999723-6309f972-9e12-11e8-9c54-60ec9076bf80.png)


As you can see, I have an almost monochrome panel. (I don't even color highlight the error type.) 

Maybe:

![image](https://user-images.githubusercontent.com/8558/43999838-d7f14cf2-9e14-11e8-8b60-40e7a30e00f0.png)

With custom colors:

![image](https://user-images.githubusercontent.com/8558/44000029-87bf57ca-9e18-11e8-8f6f-2b2e63bf5d51.png)

![image](https://user-images.githubusercontent.com/8558/44000033-9314184a-9e18-11e8-8348-48ec9fc5be01.png)

Again yours:

![image](https://user-images.githubusercontent.com/8558/44000052-db467734-9e18-11e8-8eeb-7581442bf0a9.png)

![image](https://user-images.githubusercontent.com/8558/44000058-e6b79a3a-9e18-11e8-97d8-5f764ac4dd92.png)


I think I would go without the ':'. It looks too close.

kaste commented:
Looking at it again, I think I prefer the ':'. 🤷‍♂️

kaste commented:
It is good to go, but the scopes here just suck. `storage.type` is probably good, but `comment.line` for the linter name?

Unfortunately, I don't know enough about all of this, but shouldn't this be 
```
storage.type.linter.linter_name.sublime-linter
storage.type.linter.error_type

```

kaste commented:
Can you group the `flake8:E302` tokens, so that I can match the ':' as different from the ':' in the 'line:col' token?

Can you give the message a scope?

Can you actually mark lines as errors/warnings so that e.g. warnings can be a bit desaturated?

braver commented:
@kaste try it with the last commit!

kaste commented:
That's really interesting. While playing with it, I changed 

```
meta.linter.error.body.sublime_linter
meta.linter.body.error.sublime_linter

(same for warnings)
```

so that I can match `meta.linter.body` to match both.

Also: Shouldn't we use `entity.type.linter_name` -> `entity.type.linter-name`? Currently, we have `line-number` and `col-number`, but `linter_name` and `error_code`.

braver commented:
> Shouldn't we use entity.type.linter_name -> entity.type.linter-name? Currently, we have line-number and col-number, but linter_name and error_code.

Indeed!

braver commented on panel/panel.sublime-syntax:
```diff
@@ -37,13 +44,21 @@ contexts:
     - include: pop-at-end
     - match: \berror\b
-      scope: markup.error.sublime_linter
+      scope: meta.linter.error-type.sublime_linter markup.error.sublime_linter
```
the meta is actually usually intended to indicate you're matching a group of other, more specific scopes: https://www.sublimetext.com/docs/3/scope_naming.html#meta

> Meta scopes are used to scope larger sections of code or markup, generally containing multiple, more specific scopes.

kaste replied:
Suggestion? 

Usually you would go for `x.linter.problem.error`, `x.linter.problem.warning`, and for other types just `x.linter.problem`. 

I really don't know why we use `entity.type.linter-name` but `storage.type.error-code`. Why do we change the namespace from _entity_ to _storage_?

braver replied:
> Why do we change the namespace from entity to storage?

To differentiate them in a way that will result in a different color in most color schemes. *No* color schemes, now or in the future, will have specified something for `linter-name`. So if they're both just `entity.type` both will look exactly the same. There should be something to differentiate them in a way that makes sense semantically, and there should be something in there that helps existing color schemes assign colors.

braver replied:
I could go with `entity.name.class` for linter-name, but the default color schemes don't have anything for `entity.name.label` so the code would look just like the name. Which isn't bad, but I like some differentiation there. `entity.name.function` would work though. `entity.name.tag` would work for the error types, but we already have `markup.error/warning` there ensuring that we get signal colors by default.

kaste replied:
That's not what I'm seeing. 

![image](https://user-images.githubusercontent.com/8558/44345051-79c60080-a492-11e8-81bd-a90f4e26ea8a.png)

Most default schemes actually **don't** have anything for `*.class`, but for `label` and for `tag`. (Basically except 'Celeste')

**None** of the default schemes define `markup.error` or `markup.warning`. Here I wanted the scope to be `entity.name.tag markup.error`, in case someone has that defined.

Now, I think for our purpose class, label, tag can be used interchangeably. So, we have at most 3 different default colors, we can choose from. We can use all three or make it less Kindergarten. E.g. `entity.name.label.error-type`, `entity.name.label.linter-name` and `entity.name.label.error-code`

braver replied:
> None of the default schemes define markup.error or markup.warning

Dang, you're right!? What crap editor is this anyway?

Ok, so I won't really get the control that I want. Let's keep it simple then.

We do have the colorish scopes to get the signal colors for error/warning like we do elsewhere. Let's use the same scopes there as our defaults: `region.yellowish markup.warning.sublime_linter` and `region.redish markup.error.sublime_linter`. I don't believe we need an additional `entity.name` scope there because there are already plenty of hooks to customize it if you want. 

For the others let's use `entity.name.label.linter-name` and `entity.name.label.error-code`. I'd like to have the differentiation, but we don't have to color all the things.

braver replied:
Oh, right, we can't use the colorish scopes because they also do the background color. Dang.

braver replied:
Ok, you've got the overall idea and I'm apparently lacking the sharpness to really add anything of use here. Final thing I'd like to say is that I think the error-type should have a really different scope than the linter:error block so it doesn't all look the same in all the default color schemes. Beyond that I make my own color schemes anyway.

kaste commented:
Ah, well, it's git and we can still take whatever we want. ...

- Match the actual error message with `meta.linter.message`
- Ensure compatibility with any error type (related #1475) (via `meta.linter.error-type`)
- Added tests for fun

kaste commented:
What I don't get to work. error_code is optional, and I would like to omit the ':' if there is no code. Otherwise:

![image](https://user-images.githubusercontent.com/8558/44116095-db7d5c1a-a00f-11e8-97a8-002a69427a41.png)

which is totally ugly.

braver commented:
> error_code is optional, and I would like to omit the ':' if there is no code. 

The `:` is baked into [the template](https://github.com/SublimeLinter/SublimeLinter/blob/master/panel_view.py#L351). We'd have to make that more flexible.

kaste commented:
> The : is baked into the template.

Sure 🙄 I changed that, but I could not get the syntax highlighting right after that.

kaste commented:
It's in `expect-linter-type:`. I had no luck playing with `- match: (\S+)(:)(\S*)`.

braver commented:
Oh right. I’ll play with it a bit later/tomorrow

braver commented:
@kaste to make the report without `:` work I merged the type and message sections because more of the line has to be taken into consideration to make the regex work. I updated the syntax and added a test. Can you try if it works with your dynamic message template?

braver commented:
Alright, @kaste let's ship it?
